File: src/sandbox/rng.ts
/**
 * A tiny deterministic PRNG so sandbox fixtures are reproducible. Seeded from the
 * VUA plus an optional run seed, never from wall-clock or Math.random, so the same
 * inputs always produce the same accounts and transactions across machines and CI.
 */

/** FNV-1a 32-bit hash, used to turn a string seed into a numeric one. */
export function hashSeed(input: string): number {
  let hash = 0x811c9dc5;
  for (let index = 0; index < input.length; index++) {
    hash ^= input.charCodeAt(index);
    hash = Math.imul(hash, 0x01000193);
  }
  return hash >>> 0;
}

export interface Rng {
  next(): number;
  int(minInclusive: number, maxInclusive: number): number;
  pick<T>(items: readonly T[]): T;
}

/** mulberry32: fast, deterministic, good enough for fixtures (not for crypto). */
export function createRng(seed: number): Rng {
  let state = seed >>> 0;
  const next = (): number => {
    state = (state + 0x6d2b79f5) >>> 0;
    let mixed = state;
    mixed = Math.imul(mixed ^ (mixed >>> 15), mixed | 1);
    mixed ^= mixed + Math.imul(mixed ^ (mixed >>> 7), mixed | 61);
    return ((mixed ^ (mixed >>> 14)) >>> 0) / 4294967296;
  };
  return {
    next,
    int(minInclusive, maxInclusive) {
      return minInclusive + Math.floor(next() * (maxInclusive - minInclusive + 1));
    },
    pick(items) {
      if (items.length === 0) {
        throw new Error('cannot pick from an empty list');
      }
      return items[this.int(0, items.length - 1)] as (typeof items)[number];
    },
  };
}

File: python/src/sahaj/sandbox/rng.py
"""A tiny deterministic PRNG so sandbox fixtures are reproducible. Seeded from the
VUA plus an optional run seed, never from the wall clock or a global random source,
so the same inputs always produce the same accounts and transactions.
"""

from __future__ import annotations

from typing import Sequence, TypeVar

_MASK = 0xFFFFFFFF
T = TypeVar("T")


def hash_seed(text: str) -> int:
    """FNV-1a 32-bit hash, used to turn a string seed into a numeric one."""
    value = 0x811C9DC5
    for char in text:
        value ^= ord(char)
        value = (value * 0x01000193) & _MASK
    return value


class Rng:
    """mulberry32: fast, deterministic, good enough for fixtures (not for crypto)."""

    def __init__(self, seed: int) -> None:
        self._state = seed & _MASK

    def next(self) -> float:
        self._state = (self._state + 0x6D2B79F5) & _MASK
        mixed = self._state
        mixed = ((mixed ^ (mixed >> 15)) * (mixed | 1)) & _MASK
        mixed ^= (mixed + (((mixed ^ (mixed >> 7)) * (mixed | 61)) & _MASK)) & _MASK
        mixed &= _MASK
        return ((mixed ^ (mixed >> 14)) & _MASK) / 4294967296

    def int(self, min_inclusive: int, max_inclusive: int) -> int:
        return min_inclusive + int(self.next() * (max_inclusive - min_inclusive + 1))

    def pick(self, items: Sequence[T]) -> T:
        if not items:
            raise ValueError("cannot pick from an empty list")
        return items[self.int(0, len(items) - 1)]
